Add JGF reader support

dongahn opened the pull request:
This PR adds JGF reader support. This should not be merged before PR #520. 

- Introduce reader classes into the new `resource/readers` directory and refactor `resource_graph_db_t` into its own directory (`resource/store`). Reader classes not only refactor the existing reader code (i.e., GRUG and Hwloc) but also add the new JGF reader code.

- Integrate the new reader support into `resource-query` and `resource` module.

- As part of that, change the command line options for `resource-query` and the load options for `resource` module. For `resource-query`, `--grug` and `--hwloc-xml` are deprecated and `--load-file` and `--load-format` are added instead. `--hwloc-whitelist` is now changed to `--load-whitelist`. For the `resource` module, `grug-conf`, `hwloc-xml`, and `hwloc-whitelist` have been deprecated. Instead, `load-file`, `load-format`, and `load-whitelist` are used.

- Adjust the existing tests for these option name changes.

- Add three new test cases to ensure the correctness of the JGF reader.

dongahn commented:
`t3014-resource-basic-jgf.t` tests are failing in the Travis CI. Let me look.

dongahn commented:
Travis is green now. @SteVwonder, @tpatki and @milroy: this PR should also be ready for your reviews. This should just be merged after PR #520, though.

dongahn commented:
PR #523 has been merged. This needs to be rebased to the upstream master.

dongahn commented:
Rebased to the current master.

dongahn commented:
Just rebased to the master with PR #520 and forced a push.

tpatki commented on resource/modules/resource_match.cpp:
```diff
@@ -53,9 +53,9 @@ using namespace Flux::resource_model;
-    string hwloc_whitelist;
+    string load_file;              /* load file name */
+    string load_format;            /* load reader format */
+    string load_whitelist;         /* load resource whitelist */
```
@dongahn: 
Shall provide a detailed review soon, but had a quick comment.

Can we change the variable names to reflect that this is for a graph? Maybe something like `graph_input_file`, `graph_format` is better than `load_file` and `load_format`? Also, is the `whitelist` specific to `hwloc`? Do other formats (GRUG/JGF) use the whitelist?

dongahn replied:
The variable names are matched with the command line options. There other graph related option in resource query so I didn't use graph. If you feel strongly about these, I will be happy to change them, though. Let me know.

Yes there is a good chance that white list can be supported by other readers and I made this a method in the base class for future extension.

tpatki commented on resource/modules/resource_match.cpp:
```diff
@@ -337,97 +345,129 @@ static json_t *get_string_blocking (flux_t *h, const char *key)
     if (ctx->args.reserve_vtx_vec != 0)
         ctx->db.resource_graph.m_vertices.reserve (ctx->args.reserve_vtx_vec);
+    if ( (rd = create_resource_reader (ctx->args.load_format)) == nullptr) {
+        flux_log (ctx->h, LOG_ERR, "creating load reader");
```
Shouldn't the error message be "Can't create load reader"?

Also, I am wondering if a clear set of parentheses around the `create_resource_reader (ctx->args.load_format)) == nullptr` check will make this more readable. Readability of `return_val = function () == nullptr` format is confusing.

dongahn replied:
If you are talking about the extra whitespace in `if ( (rd =`, this is a pretty common practive. In fact the famous Steven book on "Unix system programming" recommends this. If this was defined in our style guide, I can change it. But I don't recall if I have seen this rewuirement.

dongahn replied:
> Shouldn't the error message be "Can't create load reader"?

This was intentional as `LOG_ERR` generates "err" in the message prefix. But I can see why your message makes more sense. So I will change it. Thanks.

dongahn replied:
> Also, a clear set of parentheses around the create_resource_reader (ctx->args.load_format)) == nullptr before assigning to rd will make the if statement more readable. Readability of if (return_val = function () == nullptr) format is confusing.

Hmmm yes that's why I had the extra parenthesis before rc and the end of the method call. Maybe I am missing something.

tpatki commented on resource/modules/resource_match.cpp:
```diff
@@ -337,97 +345,129 @@ static json_t *get_string_blocking (flux_t *h, const char *key)
+        }
+        o = get_string_blocking (h, k);
+        hwloc_xml = json_string_value (o);
+        if ( (rc = ctx->db.load (hwloc_xml, rd, v, rank)) < 0) {
```
Minor nit: Extra space between parentheses.

dongahn replied:
See above.

tpatki commented on resource/modules/resource_match.cpp:
```diff
@@ -337,97 +345,129 @@ static json_t *get_string_blocking (flux_t *h, const char *key)
+    }
+    buffer << in_file.rdbuf ();
+    in_file.close ();
+    if ( (rc = ctx->db.load (buffer.str (), rd)) < 0) {
```
Not sure if this is part of the style guide, but there's an extra space between the parenthesis here.

dongahn replied:
Please see above.

tpatki commented on resource/modules/resource_match.cpp:
```diff
@@ -337,97 +345,129 @@ static json_t *get_string_blocking (flux_t *h, const char *key)
+    }
+    o = get_string_blocking (h, k);
+    hwloc_xml = json_string_value (o);
+    if ( (rc = ctx->db.load (hwloc_xml, rd, rank)) < 0) {
```
Space between parentheses.

dongahn replied:
Same.

tpatki commented on resource/modules/resource_match.cpp:
```diff
@@ -337,97 +345,129 @@ static json_t *get_string_blocking (flux_t *h, const char *key)
+        if (rd->set_whitelist (ctx->args.load_whitelist) < 0)
+            flux_log (ctx->h, LOG_ERR, "setting whitelist");
+        if (!rd->is_whitelist_supported ())
+            flux_log (ctx->h, LOG_WARNING, "whitelist unsupported");
```
Not sure if whitelisting should be a part of the base class (see earlier comment on whether we expect future graph formats other than hwloc to have whitelists).

dongahn replied:
Yes exactly. This is there in the base class for future extension.

tpatki commented on resource/modules/resource_match.cpp:
```diff
@@ -337,97 +345,129 @@ static json_t *get_string_blocking (flux_t *h, const char *key)
-            rc = -1;
-            flux_log (ctx->h, LOG_ERR, "error in generating resources");
+        if (populate_resource_db_kvs (ctx, rd) < 0) {
+            flux_log (ctx->h, LOG_ERR, "loading resources from the KVS");
```
Maybe change the message to "Error loading resources from KVS"?

dongahn replied:
My understanding was "err" will be emitted with `LOG_ERR` and this was part of my effort to avoid redundancy. Let me double check though.

tpatki commented on resource/readers/resource_reader_grug.cpp:
```diff
@@ -0,0 +1,465 @@
+    int get_rank ();
+
+private:
+    vtx_t emit_vertex (ggv_t u, gge_t e, const gg_t &recipe,
```
Change `ggv_t` to `vtx_t`?

dongahn replied:
Actually here v and e are an vertex and edge of the resource generation graph. So `ggv_t` should be correct I think.

tpatki commented on resource/readers/resource_reader_grug.cpp:
```diff
@@ -0,0 +1,465 @@
+    ~dfs_emitter_t ();
+
+    void tree_edge (gge_t e, const gg_t &recipe);
+    void finish_vertex (ggv_t u, const gg_t &recipe);
```
Change `ggv_t` to `vtx_t`?

dongahn replied:
Same as above.

tpatki commented on resource/readers/resource_reader_grug.cpp:
```diff
@@ -0,0 +1,465 @@
+    int gen_id (gge_t e, const gg_t &recipe,
+                int i, int sz, int j);
+
+    map<ggv_t, vector<vtx_t>> m_gen_src_vtx;
```
Change `ggv_t` to `vtx_t`?

dongahn replied:
Same.

tpatki commented on resource/readers/resource_reader_grug.cpp:
```diff
@@ -0,0 +1,465 @@
+    g[e].name[recipe[ge].e_subsystem] = recipe[ge].rrelation;
+}
+
+vtx_t dfs_emitter_t::emit_vertex (ggv_t u, gge_t e, const gg_t &recipe,
```
`ggv_t` to `vtx_t`

dongahn replied:
Ditto.

tpatki commented on resource/readers/resource_reader_grug.cpp:
```diff
@@ -0,0 +1,465 @@
+    int id = 0;
+
+    if (src_v == boost::graph_traits<resource_graph_t>::null_vertex ()) {
+        // ROOT!!
```
Minor: maybe change the comment to be `// Root vertex of graph.` as opposed to `ROOT!!`

dongahn replied:
Sure. I will change it.

tpatki commented on resource/readers/resource_reader_grug.cpp:
```diff
@@ -0,0 +1,465 @@
+{
+    vtx_t src_vtx, tgt_vtx;
+    ggv_t src_ggv = source (e, recipe);
+    ggv_t tgt_ggv = target (e, recipe);
```
I'm unsure about legitimate use of ggv_t versus vtx_t. Will revisit these once I understand this more clearly.

dongahn replied:
I left a comment to an issue ticket where you asked this question. Thanks

tpatki commented on resource/utilities/resource-query.cpp:
```diff
@@ -466,52 +468,51 @@ static void control_loop (resource_context_t *ctx)
-        ctx->params.grug = "conf/default";
-
-    gettimeofday (&st, NULL);
+    if ( (rd = create_resource_reader (ctx->params.load_format)) == nullptr) {
```
Space between parenthesis

dongahn replied:
Same as above.

tpatki commented on resource/utilities/resource-query.cpp:
```diff
@@ -466,52 +468,51 @@ static void control_loop (resource_context_t *ctx)
+    in_file.close ();
 
+    gettimeofday (&st, NULL);
+    if ( (rc = ctx->db.load (buffer.str (), rd)) != 0) {
```
Space between parenthesis.

dongahn replied:
Same.

tpatki commented:
@dongahn: I made a few very minor comments, I will have to look at this in detail (esp the new JGF reader and tests) and I shall get to it soon.

tpatki commented:
Also, this will need to be rebased. t3014 and t4007 were part of the variation-aware scheduler that was just merged, and those test cases will have to be renamed to a different number.

dongahn commented:
Thanks @tpatki. I left my responses above. I will change a few things you spotted and as part of this I will adjust the test numbering that conflicts with your variation PR.

dongahn commented:
I just forced a push. I rebased this to the latest master with @tpatki's variation aware scheduling work. And I believe I have also addressed all the issues identified from the review discussions thus far. 

@SteVwonder and/or @milroy : do you have other reviews?

milroy commented:
@dongahn I am reviewing this PR now and will be finished with it today.

milroy commented on resource/readers/resource_reader_hwloc.cpp:
```diff
@@ -0,0 +1,369 @@
+             */
+            if (strncmp(obj->name, "cuda", 4) == 0)
+                id = atoi (obj->name + 4);
+            else if (strncmp(obj->name, "opencl", 6) == 0) {
```
Nits: spaces after `strncmp`.

dongahn replied:
Will do.

milroy commented on resource/readers/resource_reader_jgf.cpp:
```diff
@@ -0,0 +1,349 @@
+    for (auto kv : g[v].paths) {
+        g[v].idata.member_of[kv.first] = "*";
+        m.by_path[kv.second].push_back (v);
+        if (std::count(kv.second.begin(), kv.second.end(), '/') == 1)
```
Spaces after `begin` and `end`.

dongahn replied:
Thanks. (Maybe we need some format enforcer :-).

dongahn commented on resource/store/resource_graph_store.hpp:
```diff
@@ -0,0 +1,97 @@
+    std::map<subsystem_t, vtx_t> roots;
+    std::map<std::string, std::vector <vtx_t>> by_type;
+    std::map<std::string, std::vector <vtx_t>> by_name;
+    std::map<std::string, std::vector <vtx_t>> by_path;
```
Agreed! The changes like this are now lumped into either hardening and/or optimization project in my kanban board.

dongahn commented on resource/readers/resource_reader_grug.cpp:
```diff
@@ -0,0 +1,465 @@
+/*****************************************************************************\
+ *  Copyright (c) 2014 Lawrence Livermore National Security, LLC.  Produced at
```
I don't know what is the conversion when changes are made across multiple years. Like 2014 - 2019. I will change it that way and see how it looks.

dongahn commented on resource/readers/resource_reader_grug.cpp:
```diff
@@ -0,0 +1,465 @@
+        if (target (*ei, g) == tgt_v) {
+            e = (*ei);
+            return e;
+        }
```
Yes, once you find the matching edge, you don't need to further traverse out edges so you return. 

As we discussed before, being able to find an edge with O(map) like performance will greatly be helpful.

milroy replied:
Yes, that makes sense.  I've higher performance lookups in my WIP on issue #511.

dongahn replied:
Great! You may want to split your WIP into multiple PRs and get some early reviews before it gets too large. It seems a faster edge lookup as a standalone topic can make a good enough PR. Also there are some performance benchmarks you can try your lookup logic to see how much it improves the performance.

dongahn commented on resource/readers/resource_reader_grug.cpp:
```diff
@@ -0,0 +1,465 @@
+
+            src_vtx = *src_it;
+            for (tgt_it = m.by_type[recipe[tgt_ggv].type].begin();
+                 tgt_it != m.by_type[recipe[tgt_ggv].type].end(); tgt_it++) {
```
Good catch! I will fix them.

dongahn commented on resource/readers/resource_reader_grug.cpp:
```diff
@@ -0,0 +1,465 @@
+
+            src_vtx = *src_it;
+            for (tgt_it = m.by_type[recipe[tgt_ggv].type].begin();
+                 tgt_it != m.by_type[recipe[tgt_ggv].type].end(); tgt_it++) {
```
Ditto!

dongahn commented on resource/readers/resource_spec_grug.cpp:
```diff
@@ -0,0 +1,221 @@
+    for (i=0; str2genmeth[i].str != ""; ++i)
+        if (str2genmeth[i].str == s)
+            return (gen_meth_t)str2genmeth[i].e;
+    return (gen_meth_t)str2genmeth[i].e;
```
I don't think we have that required by our coding guide. My preference is not to have a space for casting.

dongahn commented on resource/readers/resource_spec_grug.cpp:
```diff
@@ -0,0 +1,221 @@
+        vtx_subsystem_map_t v_ss_map = get(&resource_pool_gen_t::subsystem, g);
+        edg_relation_map_t e_rel_map = get(&relation_gen_t::relation, g);
+        edg_gen_method_map_t e_gen_map = get(&relation_gen_t::gen_method, g);
+        edg_multi_scale_map_t e_ms_map = get(&relation_gen_t::multi_scale, g);
```
Good catch. I will fix those.

dongahn commented:
Thanks @milroy! I will address the issues above and push my changes.

milroy commented on resource/readers/resource_reader_base.hpp:
```diff
@@ -0,0 +1,102 @@
+     * \return       0 on success; non-zero integer on an error
+     */
+    virtual int unpack_at (resource_graph_t &g, resource_graph_metadata_t &m,
+                           vtx_t &vtx, const std::string &str, int rank = -1) = 0;
```
`param vtx` will have edges to and/or from each of the subsystem roots of the unpacked resource graph, correct?  Also, is `vtx` in the original graph but not in the unpacked graph?

dongahn replied:
@milroy: great question. Easy one first:

> Also, is vtx in the original graph but not in the unpacked graph?

Yes, it is a vertex that should only be in the original graph. This interface is currently only wired with the hwloc reader to attach a new set of subgraph created by each of the ranked hwloc xml objects into the root vertex: the `cluster` vertex. 

> param vtx will have edges to and/or from each of the subsystem roots of the unpacked resource graph, correct?

The intent is broader than this.

Say, the unpacked graph is  `node10->core0`. While `node10` is the root of the subgraph, it is not the root of the overall graph. I detect the root of the subgraph by parsing each vertex's paths field. If `node10` vertex's paths has `containment: /cluster/node10`, it won't be identified as the root of the containment subsystem. 

Also if the unpacked graph is  `node10->core0, node20->core30`, I expect that this interface attach both `node10` and `node20` to the `vtx`.  (We don't currently have this use case as hwloc reader reads one rank at a time.) If this were to work, it is almost like the caller should specify what those vertices are, which need to be attached to `vtx`. ..

If you want to propose something more general than this interface, which can help your work, maybe we can work on an interface revision after this PR lands?

milroy replied:
> This interface is currently only wired with the hwloc reader to attach a new set of subgraph created by each of the ranked hwloc xml objects into the root vertex: the cluster vertex.

But the intention is that eventually `vtx` can be any vertex in the original graph?

> While node10 is the root of the subgraph, it is not the root of the overall graph. I detect the root of the subgraph by parsing each vertex's paths field.

That's sensible.  I was abusing the term "subsystem roots" as referring to the top-level node(s) in the unpacked graph (the subgraph).  Can the subgraph have multiple roots (i.e. the subgraph contains vertices in multiple subsystems)? 

> Also if the unpacked graph is node10->core0, node20->core30, I expect that this interface attach both node10 and node20 to the vtx. (We don't currently have this use case as hwloc reader reads one rank at a time.) If this were to work, it is almost like the caller should specify what those vertices are, which need to be attached to vtx. ..

This is my main curiosity.  While the path to vertex mapping is one-to-one, the vertex to path mapping is many-to-one: `g[v].paths[subsys] = prefix + "/" + g[v].name;`.  This suggests to me that node10 could be attached to one `vtx` from each `subsys` where `g[node10].paths[subsys]` is defined.

> If you want to propose something more general than this interface, which can help your work, maybe we can work on an interface revision after this PR lands?

Yes, I think we will need a more general interface which accepts a node to vertex map.  That work is definitely out of scope for this PR.

SteVwonder replied:
> Yes, I think we will need a more general interface which accepts a node to vertex map. I agree, that work is out of scope for this PR.

Do you want to track the idea with a new issue?

milroy commented:
@SteVwonder, @dongahn, and I decided that the final review isn't necessary.  I'm merging this PR.

milroy commented:
@dongahn it looks like you still need to add spaces after a few function calls.  After that I can merge this PR, per our discussion with @SteVwonder.

dongahn commented:
> @dongahn it looks like you still need to add spaces after a few function calls. After that I can merge this PR, per our discussion with @SteVwonder.

Yes, I planned to do this after some urgent tasks. Stay tuned.

milroy commented:
@dongahn looks good!  The copyright year range is a good solution.

dongahn commented:
@milroy: I squashed all the intermediate commits. Once Travis turns green, please go ahead and merge. Thank you @milroy, @tpatki and @SteVwonder for the review.
